source/simulator/melt.cc

```diff
@@ -1234,6 +1333,272 @@ namespace aspect
+      {}
+
+      FEValues<dim>                            finite_element_values;
+
+      std::vector<types::global_dof_index>     local_dof_indices;
+      std::vector<unsigned int>                dof_component_indices;
+
+      MaterialModel::MaterialModelInputs<dim>  material_model_inputs;
+      MaterialModel::MaterialModelOutputs<dim> material_model_outputs;
+
+    };
+
+    template <int dim>
+    struct PcCopyData
+    {
+      explicit PcCopyData (const unsigned int stokes_dofs_per_cell)
+      {
+        nonzero_dof_indices.reserve(stokes_dofs_per_cell);
+      }
+
+      std::vector<types::global_dof_index>   nonzero_dof_indices;
+    };
+
+    template <int dim>
+    class PcNonZeroAssembler : public SimulatorAccess<dim>
+    {
+      public:
+        PcNonZeroAssembler(IndexSet &nonzero_entries,
+                           std::vector<bool> &is_melt_cell)
+          : nonzero_entries (nonzero_entries),
+            is_melt_cell (is_melt_cell)
+        {}
+
+
+        void
+        local_assemble (const typename DoFHandler<dim>::active_cell_iterator &cell,
+                        PcAssembleData<dim> &scratch,
+                        PcCopyData<dim> &data)
+        {
+
+          data.nonzero_dof_indices.clear();
+          cell->get_dof_indices (scratch.local_dof_indices);
+
+          scratch.finite_element_values.reinit (cell);
+
+          MeltHandler<dim>::create_material_model_outputs(scratch.material_model_outputs);
+
+          const unsigned int pc_component_index = this->introspection().variable("compaction pressure").first_component_index;
+          const unsigned int pc_base_index = this->introspection().variable("compaction pressure").base_index;
+
+          Assert(dynamic_cast<const MaterialModel::MeltInterface<dim>*>(&this->get_material_model()) !=
+                 NULL, ExcMessage("Error: The current material model needs to be derived from MeltInterface to use melt transport."));
+
+          // Prepare the data structures for assembly
+          scratch.finite_element_values.reinit (cell);
+          scratch.material_model_inputs.reinit(scratch.finite_element_values,
+                                               cell,
+                                               this->introspection(),
+                                               this->get_current_linearization_point(),
+                                               true);
+
+          this->get_material_model().evaluate(scratch.material_model_inputs,
+                                              scratch.material_model_outputs);
+
+          const double p_c_scale = dynamic_cast<const MaterialModel::MeltInterface<dim>*>(&this->get_material_model())->p_c_scale(scratch.material_model_inputs,scratch.material_model_outputs,this->get_melt_handler(), true);
+          const bool is_melt_cell = (p_c_scale > 0.0);
+
+          if (is_melt_cell)
+            {
+              for (unsigned int j=0; j<this->get_fe().base_element(pc_base_index).dofs_per_cell; ++j)
```

Comment:
Add a line `data.nonzero_dof_indices.resize(this->get_fe().base_element(pc_base_index).dofs_per_cell);` outside of the loop and access `data.nonzero_dof_indices[j]` inside.